game-core/src/main/java/games/strategy/triplea/delegate/power/calculator/AirBattleDefenseCombatValue.java

```diff
@@ -0,0 +1,93 @@
+package games.strategy.triplea.delegate.power.calculator;
+
+import games.strategy.engine.data.GameData;
+import games.strategy.engine.data.TerritoryEffect;
+import games.strategy.engine.data.Unit;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Getter;
+import lombok.NonNull;
+import lombok.Value;
+import org.triplea.java.collections.IntegerMap;
+
+/** Calculates offense strength and roll for Air Battle dice */
+@Builder
+@Value
+@Getter(AccessLevel.NONE)
+class AirBattleDefenseCombatValue implements CombatValue {
+
+  @Getter(onMethod = @__({@Override}))
+  @NonNull
+  GameData gameData;
+
+  @Getter(onMethod = @__({@Override}))
+  @NonNull
+  @Builder.Default
+  Collection<Unit> friendUnits = List.of();
+
+  @Getter(onMethod = @__({@Override}))
+  @NonNull
+  @Builder.Default
+  Collection<Unit> enemyUnits = List.of();
+
+  @Getter(onMethod = @__({@Override}))
+  @NonNull
+  @Builder.Default
+  Collection<TerritoryEffect> territoryEffects = List.of();
+
+  @Override
+  public RollCalculator getRoll() {
+    return new AirBattleDefenseRoll();
+  }
+
+  @Override
+  public StrengthCalculator getStrength() {
+    return new AirBattleDefenseStrength(gameData.getDiceSides());
+  }
+
+  @Override
+  public boolean isDefending() {
+    return false;
+  }
+
+  @Override
+  public int getDiceSides(final Unit unit) {
+    return gameData.getDiceSides();
+  }
+
+  @Value
+  @AllArgsConstructor
+  static class AirBattleDefenseRoll implements RollCalculator {
+
+    @Override
+    public RollValue getRoll(final Unit unit) {
+      return RollValue.of(unit.getUnitAttachment().getDefenseRolls(unit.getOwner()));
+    }
+
+    @Override
+    public Map<Unit, IntegerMap<Unit>> getSupportGiven() {
+      return Map.of();
+    }
+  }
+
+  @Value
+  @AllArgsConstructor
+  static class AirBattleDefenseStrength implements StrengthCalculator {
+
+    int diceSides;
+
+    @Override
+    public StrengthValue getStrength(final Unit unit) {
+      return StrengthValue.of(diceSides, unit.getUnitAttachment().getAirDefense(unit.getOwner()));
```

Comment:
This API is bothersome:
```
 unit.getUnitAttachment().getAirDefense(unit.getOwner()));
```

A 'unit' should already know who it's owner is. The encapsulation around data is very weak, we have knowledge here of attachments, unit owners, the air defense type.. Any thoughts if this can be fixed?

Reply:
We could start adding methods to the `Unit` class that encapsulates this.  So something like:
```
public int getAirDefense() {
  this.getUnitAttachment().getAirDefense(this.getOwner());
}
```

It'll make the `Unit` class a lot bigger with all of these methods since it will probably end up with all of the existing methods in `UnitType` and `UnitAttachment`.

Reply:
That could make a lot of sense now. The 'roots' of TripleA were to be completely generic, a unit could have been something like a chess piece. We've removed that level of generality to focus on the grand strategy style games.

We probably should plan such a change and design it out so we don't get ourselves caught into a corner. For example, how do we handle methods that do not apply? For example, a 'getAirDefense' method for a land unit? Do we return zero, do we throw an exception? If we return zero, can we distinguish between units that do not participate in air defense vs the ones that do but actually have zero power (IE: fodder).

It's also worth planning for how many methods and attributes we will be accumulating. Perhaps the right object model is to have strategy and/or value objects for the different aspects of a Unit. For example, a 'purchasing', a 'combatStrength', a 'movementStrategy', etc.. rather than having *all* implemenation be on Unit, instead delegate its major sub-components to additional objects.